Describe image clustering in mongodb

mfojtik opened the pull request:
Working in progress ;-)

mfojtik commented:
@bparees @soltysh PTAL

bparees commented on using_images/db_images/mongodb.adoc:
```diff
@@ -134,3 +134,91 @@ The MongoDB image can be run with mounted volumes to enable persistent storage f
 * *_/var/lib/mongodb_* - This is the database directory where
 MongoDB stores database files.
+
+=== Using the MongoDB replication
```
s/the//

bparees commented on using_images/db_images/mongodb.adoc:
```diff
@@ -134,3 +134,91 @@ The MongoDB image can be run with mounted volumes to enable persistent storage f
+
+To start with MongoDB replication, you first have to define a DeploymentConfig
+that defines a replication controller which will manage the members of the
+MongoDB cluster. To tell MongoDB server that the member will be part of the
```
To tell *a* MongoDB server

bparees commented on using_images/db_images/mongodb.adoc:
```diff
@@ -134,3 +134,91 @@ The MongoDB image can be run with mounted volumes to enable persistent storage f
+a headless service, you have to set the `portalIP` field value in the service
+definition to `None`. Then you can use a DNS query to get a list of the pod IP
+addresses that represents the current endpoints for this service.
+For example, you can use the `dig` command: `dig mongodb A +short +search`.
```
I think a better way to approach documenting this all is to describe it like you did, but also say something like "the mongodb clustered service template sets this configuration here:" and then show the relevant json from the template.

bparees commented on using_images/db_images/mongodb.adoc:
```diff
@@ -134,3 +134,91 @@ The MongoDB image can be run with mounted volumes to enable persistent storage f
+For example, you can use the `dig` command: `dig mongodb A +short +search`.
+
+When the entrypoint has the IP addresses of all running MongoDB members, you can
+create a MongoDB replication set configuration where you list all member IP
```
I can create it?  how do i?  isn't this done for me?

bparees commented on using_images/db_images/mongodb.adoc:
```diff
@@ -134,3 +134,91 @@ The MongoDB image can be run with mounted volumes to enable persistent storage f
+you have to wait until the MongoDB elect the PRIMARY member of the cluster.  You
+can create a simple MongoDB script that will run a loop and poll the MongoDB
+server till the PRIMARY member appear on the list. Since you started the
+initiation on the 'service pod', this pod will be elected.
```
again this is a bit hand-wavy for me.. either show me the exact commands i need to run, or explain that it's done for me and just tell me what it's doing.

bparees commented on using_images/db_images/mongodb.adoc:
```diff
@@ -134,3 +134,91 @@ The MongoDB image can be run with mounted volumes to enable persistent storage f
+shut down. You have to define the `restartPolicy` field in the 'service pod' to
+`Never` to prevent the 'service pod' from restarting when the container exits.
+Right after the 'service pod' goes down, other members will start a new election
+and the new PRIMARY member is being elected from the running members.
```
again show how the template configures all this

bparees commented on using_images/db_images/mongodb.adoc:
```diff
@@ -134,3 +134,91 @@ The MongoDB image can be run with mounted volumes to enable persistent storage f
+other members will replicate the data to a new pod and start a new election.
+
+The full source of replication example is available here: 
+https://github.com/openshift/mongodb/tree/master/2.4/examples/replica
```
link directly to the templates directory too

bparees commented on using_images/db_images/mongodb.adoc:
```diff
@@ -134,3 +134,91 @@ The MongoDB image can be run with mounted volumes to enable persistent storage f
+https://github.com/openshift/mongodb/tree/master/2.4/examples/replica
+
+IMPORTANT: The template provided in the example above does not use persistent
+storage. When you loose all members of the replication set, your data will be
```
s/When/if/
s/loose/lose/

"If you lose all members of the replication set simultaneously, you data will be lost"  (I assume simultaneous is accurate there and it sounds better/more secure)

mfojtik replied:
I made it sound insecure intentionally as this whole template is **very** insecure if you care about the data ;-)

mfojtik commented:
@adellape I wonder if it would make sense to start a new section named 'Clustering' next to 'using_images'... There will be more docs coming about how to scale rails/django, how to create master-slave setup for MySQL, etc... Having this all part of 'Using image' might make the docs for the image quite long.

bparees commented on using_images/db_images/mongodb.adoc:
```diff
@@ -134,3 +134,91 @@ The MongoDB image can be run with mounted volumes to enable persistent storage f
+master-slave replication and automated failover. This is the MongoDB’s
+recommended replication strategy.
+
+To start with MongoDB replication, you first have to define a DeploymentConfig
```
you can break this whole thing down into subsections
(creating the service, creating the primary pod, creating the cluster)

mfojtik commented:
@bparees want wait for @adellape on whether we want to continue documenting this like I did or move it into special 'clustering' section.

adellape commented:
@mfojtik @bparees I think we should keep as much image-specific info/config steps within their respective image-specific topics as we can. I don't think it's ridiculously long currently, IMO.

And I would agree with Ben's comments about making it clear early on and throughout as to what the template takes care of for you. While reading it, I kept wondering what I had to do myself vs what was something that was taken care of by <?>, and wanting to see relevant snippets of YAML along the way. Then at the end I realized there was an example replica template available.

mfojtik commented:
@adellape @bparees thanks! hopefully I implemented all your suggestions. I broke the section into subsections and replaced 'you do something' with 'it does something' where the template/docker image do things for you. PTAL.

adellape commented on using_images/db_images/mongodb.adoc:
```diff
@@ -134,3 +134,278 @@ The MongoDB image can be run with mounted volumes to enable persistent storage f
+https://github.com/openshift/mongodb/tree/master/2.4/examples/replica
+
+IMPORTANT: The template provided in the example above does not use persistent
+storage. When you loose all members of the replication set, your data will be
```
s/loose/lose/

adellape commented on using_images/db_images/mongodb.adoc:
```diff
@@ -134,3 +134,278 @@ The MongoDB image can be run with mounted volumes to enable persistent storage f
+lost.
+
+This section describes how to start a cluster of MongoDB servers that implements
+master-slave replication and automated failover. This is the MongoDB’s
```
s/the MongoDB's/MongoDB's/

or

"This is the recommended replication strategy for MongoDB."

adellape commented on using_images/db_images/mongodb.adoc:
```diff
@@ -134,3 +134,278 @@ The MongoDB image can be run with mounted volumes to enable persistent storage f
+
+==== Creating the service pod
+
+To initialize members created by the DeploymentConfig a "service pod" has to be
```
Needs comma: "DeploymentConfig, a"

adellape commented on using_images/db_images/mongodb.adoc:
```diff
@@ -134,3 +134,278 @@ The MongoDB image can be run with mounted volumes to enable persistent storage f
+==== Creating the service pod
+
+To initialize members created by the DeploymentConfig a "service pod" has to be
+defined in the Template. This pod will start MongoDB with `initiate` argument,
```
with the `initiate` argument,

adellape commented on using_images/db_images/mongodb.adoc:
```diff
@@ -134,3 +134,278 @@ The MongoDB image can be run with mounted volumes to enable persistent storage f
+----
+{
+  "kind": "Pod",
+  "apiVersion": "v1beta3",
```
Should be v1?

adellape commented on using_images/db_images/mongodb.adoc:
```diff
@@ -134,3 +134,278 @@ The MongoDB image can be run with mounted volumes to enable persistent storage f
+
+==== Creating headless service
+
+The `initiate` argument in container specification above will make the container
```
"in the container specification"

adellape commented on using_images/db_images/mongodb.adoc:
```diff
@@ -134,3 +134,278 @@ The MongoDB image can be run with mounted volumes to enable persistent storage f
+This will tell the replication controller to create a new MongoDB member pod.
+When a new member is created, the member entrypoint will first attempt do
+discover other running members in the cluster. Then it will choose one and add
+itself to the list of members. Once the replication configuration is update, the
```
s/update/updated/

adellape commented:
Thanks @mfojtik, changes look cool, though it needs a rebase now. Added some comments inline, too.

BTW, if it's all the same, could we show YAML for the examples? Though I suppose having YAML here and then JSON in the actual mongodb_controller.json file we're linking to might be a little weird...

bparees commented on using_images/db_images/mongodb.adoc:
```diff
@@ -134,3 +134,278 @@ The MongoDB image can be run with mounted volumes to enable persistent storage f
+that defines a replication controller which will manage the members of the
+MongoDB cluster. To tell a MongoDB server that the member will be part of the
+cluster, you have to provide additional environment variables for the container
+defined in the replication controller pod template:
```
I think this paragraph should be broken apart..  you're talking about two different things:
1) the deploymentconfig that manages the cluster
2) the deploymentconfig for a member that joins the cluster

and it's not very clear as written when you're talking about which one.

bparees replied:
actually I take it back, I guess you're only talking about the member DC here.  I would put the section that describes the one-shot service/pod before this section.

bparees replied:
after reading through the rest I guess i can see the argument for explaining the main cluster before the service+one shot pod.

bparees commented on using_images/db_images/mongodb.adoc:
```diff
@@ -134,3 +134,278 @@ The MongoDB image can be run with mounted volumes to enable persistent storage f
+
+==== Creating the final replication set
+
+When the entrypoint has the IP addresses of all running MongoDB members, it
```
not clear what "when the entrypoint has the ip addresses" means... what's the entrypoint?

I assume you mean the script that is run as the image's entrypoint, but it's not obvious.

bparees commented on using_images/db_images/mongodb.adoc:
```diff
@@ -134,3 +134,278 @@ The MongoDB image can be run with mounted volumes to enable persistent storage f
+addresses and then initiate the replication set by `rs.initiate(config)`. 
+Then the script waits until the MongoDB elect the PRIMARY member of the cluster.
+
+Once the PRIMARY member has been elected, the entrypoint will start creating
```
again not obvious what "the entrypoint" means.

mfojtik commented:
@adellape yeah, I would like to keep JSON because the linked template is JSON... it might be hard for reader to convert from YAML to JSON :-)

mfojtik commented:
@adellape @bparees rebased and comments addressed. Thanks!

bparees commented:
lgtm.

adellape commented on using_images/db_images/mongodb.adoc:
````diff
@@ -163,3 +163,278 @@ The MongoDB templates should have been registered in the `openshift` project by
+```
+
+This will tell the replication controller to create a new MongoDB member pod.
+When a new member is created, the member entrypoint will first attempt do
````
s/do/to/

adellape commented:
Just the one typo inline, otherwise LGTM!
